File: 3538.merge-operations-for-minimum-travel-time.java
#
# @lc app=leetcode id=3538 lang=java
#
# [3538] Merge Operations for Minimum Travel Time
#

# @lc code=start
class Solution {
    public int minTravelTime(int l, int n, int k, int[] position, int[] time) {
        final int INF = 2000000000;
        int[][][] dp = new int[n][k + 1][101];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= k; j++) {
                for (int t = 0; t <= 100; t++) {
                    dp[i][j][t] = INF;
                }
            }
        }
        dp[0][0][time[0]] = 0;
        long[] presum = new long[n + 1];
        for (int i = 0; i < n; i++) {
            presum[i + 1] = presum[i] + time[i];
        }
        for (int i = 1; i < n; i++) {
            for (int p = 0; p < i; p++) {
                int merges_add = i - p - 1;
                if (merges_add < 0) continue;
                for (int jprev = 0; jprev <= k - merges_add; jprev++) {
                    for (int s = 0; s <= 100; s++) {
                        if (dp[p][jprev][s] == INF) continue;
                        long sumadd = presum[i] - presum[p + 1];
                        int effnew = time[i] + (int) sumadd;
                        if (effnew > 100) continue;
                        long addcost = (long) (position[i] - position[p]) * s;
                        long newcost = (long) dp[p][jprev][s] + addcost;
                        int jj = jprev + merges_add;
                        if (newcost < dp[i][jj][effnew]) {
                            dp[i][jj][effnew] = (int) newcost;
                        }
                    }
                }
            }
        }
        int ans = INF;
        for (int t = 0; t <= 100; t++) {
            ans = Math.min(ans, dp[n - 1][k][t]);
        }
        return ans;
    }
}
# @lc code=end

File: 3538.merge-operations-for-minimum-travel-time.go
#
# @lc app=leetcode id=3538 lang=golang
#
# [3538] Merge Operations for Minimum Travel Time
#

# @lc code=start
import "math"

func minTravelTime(l int, n int, k int, position []int, time []int) int {
	prefix := make([]int, n+1)
	for i := 0; i < n; i++ {
		prefix[i+1] = prefix[i] + time[i]
	}

	const INF int64 = 1e18

	dp := make([][][][]int64, n)
	for i := range dp {
		dp[i] = make([][][]int64, n)
		for j := range dp[i] {
			dp[i][j] = make([][]int64, k+1)
			for m := range dp[i][j] {
				dp[i][j][m] = make([]int64, 101)
				for ad := range dp[i][j][m] {
					dp[i][j][m][ad] = INF
				}
			}
		}
	}

	for i := 0; i < n; i++ {
		dp[i][i][0][0] = 0
	}

	for length := 2; length <= n; length++ {
		for L := 0; L <= n-length; L++ {
			R := L + length - 1
			maxMerges := k
			if R - L - 1 < maxMerges {
				maxMerges = R - L - 1
			}

			for m := 0; m <= maxMerges; m++ {
				for Q := L; Q < R; Q++ {
					rightMerges := R - Q - 1
					if rightMerges > m {
						continue
					}
					leftMerges := m - rightMerges
					dist := int64(position[R] - position[Q])
					var minCost int64 = INF
					for addedQ := 0; addedQ <= 100; addedQ++ {
						leftCost := dp[L][Q][leftMerges][addedQ]
						if leftCost == INF {
							continue
						}
						effTimeQ := int64(time[Q]) + int64(addedQ)
						cost := leftCost + effTimeQ * dist
						if cost < minCost {
							minCost = cost
						}
					}
					if minCost == INF {
						continue
					}
					addedR := prefix[R] - prefix[Q+1]
					if addedR > 100 {
						continue
					}
					if minCost < dp[L][R][m][addedR] {
						dp[L][R][m][addedR] = minCost
					}
				}
			}
		}
	}

	ans := INF
	for added := 0; added <= 100; added++ {
		if dp[0][n-1][k][added] < ans {
			ans = dp[0][n-1][k][added]
		}
	}

	return int(ans)
}
# @lc code=end